### ISSUE TITLE
Twisted should use sdist to build tarballs instead

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7985]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#7985|
|Type|enhancement|
|Created|2015-08-04 04:27:23Z|
|Branch|https://github.com/twisted/twisted/tree/sdist-use-7985-4|

<details><summary>Searchable metadata</summary>

```
trac-id__7985 7985
type__enhancement enhancement
reporter__hawkowl hawkowl
priority__normal normal
milestone__ 
branch__branches_sdist_use_7985_4 branches/sdist-use-7985-4
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__release_management release management
keywords__None None
time__1438662443726499 1438662443726499
changetime__1457407295816758 1457407295816758
version__None None
owner__hawkowl hawkowl
cc__radix
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

(In [45390]) Branching to sdist-use-7985.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

(In [45774]) Branching to sdist-use-7985-2.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl commented|
|-|-|

This makes the relevant changes.

To test this, you will need to make an sdist (```python setup.py sdist```) and compare the files in it to a released tarball. Extract the released one to released, the sdist to sdist. I then compared the files with ```cd released && find . > ../released.txt && cd ../sdist && find . > ../sdist.txt && cd .. && comm released.txt sdist.txt``` and inspecting it.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

What is the purpose of this change?  Is sdist a moral good in and of itself?

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl commented|
|-|-|

sdist is the moraliest good.

But, in seriousness, this gives us:

- Standardisation with other Python projects, in the realm of packaging. Making a tarball is now the same as every other project.

- sdist does some metadata creation that lets our tarballs be uploaded by ```twine```, so it means we don't have to reinvent the ```PKG-INFO``` creator (which I don't even know where we'd start), and we can automate our releases a little bit more.

- We remove a bunch of stuff that existed because sdist wasn't good, I'm assuming, and now it's better. The code does not need to exist now.

-hawkie

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban set owner to @hawkowl|
|-|-|

```
recursive-include twisted *.misc *.bugfix *.doc *.feature *.removal
```

do we need these files in the release?
I was expecting that the release will remove all these files and we will not release code with those files.

Why are those files ``Topfiles`` ?

-----

Do we need to distribute tox.ini and _preamble.py with the sdist archive?

---

Is this line ok ? It looks like mixin module names with path names.

```
args['py_modules'] = modulesToInstall + testDataFiles
```

or we should rename testDataFiles to testDataModules

It has no news file fragment.

--------

This branch removes a couple of tests but no new tests are added.
Does the existing tests provide covergate for the ``setup.py sdist`` changes?

Does this branch comes with the associated documentation which instruc what should we use not for release instead of the removed code?


Please check my comments and resubmit.

I am happy to see that we can use the ``setup.py sdist`` but I think that it needs at least a few changes in the documentation.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl commented|
|-|-|

[#7759](https://github.com/twisted/twisted/issues/7759) would be fixed by this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

(In [46888]) Branching to sdist-use-7985-3.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_10"></a>|@hawkowl removed owner|
|-|-|

Builds look good, I'm happy. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl commented|
|-|-|

https://buildbot.twistedmatrix.com/builders/check-manifest/builds/105 has the check-manifest build, too.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban commented|
|-|-|

Major comments:

1. I assume the plan is to run the tests suite from the sdist and in this way also test the whole sdist thing. This is not documented. Maybe MANIFEST.in should start with a comment documenting what files should be included here and based on which rules.

2. The replacement for bin/admin/build-tarballs is not documented ... and the removed command is still documented at https://twistedmatrix.com/trac/wiki/ReleaseProcess

3. Please include a news fragment. I would say that it should have a .feature one advertising package developers that the distribution script were change.

Minor comments

1. Instead of 'PickyBuildPy' maybe we can have 'BuildOnlyPortedModules'. Same for PickyBuildScripts... which would be nice to have a docstring describing its purpose.

2. I would say something similar to DistributionBuilderTests.test_twistedDistribution should have been part of the new patch... but I hope that by integrating sdist into our continuous tests we can cover these changes in this way.

------

Looking forward for having tox as our buildslave driver and put the changes from tox.ini under continuous testing.

Please check my comments and merge.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_13"></a>|@hawkowl commented|
|-|-|

Hi Adi, thanks for the review.

I've re-opened, done the work, and put up [[#4543](https://github.com/twisted/twisted/issues/4543)](https://github.com/twisted/twisted/issues/4543) for review, which will add the release process to version control. Once #4543 is merged, I will update this ticket with the updated Twisted release process, put it up for review, and then both the process and the implementation can be merged without the other being out of date.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_14"></a>|@hawkowl commented|
|-|-|

(In [46932]) Branching to sdist-use-7985-4.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_15"></a>|@hawkowl commented|
|-|-|

I've added:

- Comments about the MANIFEST.in

- Documented the replacement of build-tarballs in the release process

- Added a news fragment

As per testing the sdist elements, we have a check-manifest builder which will go green after this and perform the same duty as the old tests did, and (very) soon we will move to installed testing in tox, where the setup will be put under continuous test.

Builders spun. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_16"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Looks good. Thanks!

I guess that to really be able to review/test all this, together with the documentation it would require to have a different release manager for on release.

I trust you that the information is accurate :) so please merge as we don't need to make it more complicated :)

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_17"></a>|@hawkowl set status to `closed`|
|-|-|

(In [46942]) Merge sdist-use-7985-4: Use setup.py sdist to build tarballs

Author: hawkowl
Reviewer: adiroiban
Fixes: [#7985](https://github.com/twisted/twisted/issues/7985)
